src/rank_llm/rerank/rankllm.py

```diff
@@ -192,6 +239,59 @@ def sliding_windows(
             start_pos = start_pos - step
         return rerank_result
 
+    def sliding_windows_batched(
+        self,
+        retrieved_results: List[Result],
+        rank_start: int,
+        rank_end: int,
+        window_size: int,
+        step: int,
+        shuffle_candidates: bool = False,
+        logging: bool = False,
+    ) -> List[Result]:
+        """
+        Applies the sliding window algorithm to the reranking process for a batch of result objects.
+
+        Args:
+            retrieved_results (List[Result]): The list of result objects to process.
+            rank_start (int): The start index for ranking.
+            rank_end (int): The end index for ranking.
+            window_size (int): The size of each sliding window.
+            step (int): The step size for moving the window.
+            shuffle_candidates (bool, optional): Flag to shuffle candidates before processing. Defaults to False.
+            logging (bool, optional): Flag to enable logging of operations. Defaults to False.
+
+        Returns:
+            List[Result]: The list of result objects after applying the sliding window technique.
+        """
+        rerank_results = [copy.deepcopy(result) for result in retrieved_results]
+        if shuffle_candidates:
+            for rerank_result in rerank_results:
+                # Shuffle rerank_result hits between rank_start and rank_end
+                rerank_result.hits[rank_start:rank_end] = random.sample(
+                    rerank_result.hits[rank_start:rank_end],
+                    len(rerank_result.hits[rank_start:rank_end]),
+                )
+                # Rescore all candidates with 1/rank
+                for i, hit in enumerate(rerank_result.hits):
+                    hit["score"] = 1.0 / (i + 1)
+                    hit["rank"] = i + 1
+
+        end_pos = rank_end
+        start_pos = rank_end - window_size
+
+        # end_pos > rank_start ensures that the list is non-empty while allowing last window to be smaller than window_size
+        # start_pos + step != rank_start prevents processing of redundant windows (e.g. 0-20, followed by 0-10)
+        while end_pos > rank_start and start_pos + step != rank_start:
+            print(f"start_pos: {start_pos}, end_pos: {end_pos}")
+            start_pos = max(start_pos, rank_start)
+            rerank_results = self.permutation_pipeline_batched(
```

Comment:
looks like the logic is identical to the old sliding window except calling batch permutation pipeline vs non-batch? if this is the case, please merge the two functions and use a condition to call the proper function, rather than copy pasting the whole function